src/bidiMapper/scripts/eval.es

```diff
@@ -0,0 +1,115 @@
+(() => {
+  class Serializer {
+    _idToObject = new Map();
+    _objectToId = new WeakMap();
+
+    constructor() {}
+
+    serialize(value) {
+      if (value === undefined) {
+        return { type: 'undefined' };
+      } else if (value === null) {
+        return { type: 'null' };
+      } else if (typeof value === 'string') {
+        return { type: 'string', value };
+      } else if (typeof value === 'number') {
+        let serialized;
+        if (Number.isNaN(value)) {
+          serialized = 'NaN';
+        } else if (value === 0 && 1 / value === -Infinity) {
+          serialized = '-0';
+        } else if (value === -Infinity) {
+          serialized = '-Infinity';
+        } else if (value === +Infinity) {
+          serialized = '+Infinity';
+        } else {
+          serialized = value;
+        }
+        return { type: 'number', value: serialized };
+      } else if (typeof value === 'boolean') {
+        return { type: 'boolean', value };
+      } else if (value instanceof Object) {
+        // TODO: Recursive serialize.
+        let id = this._objectToId.get(value);
+        if (!id) {
+          id = uuid();
+          this._objectToId.set(value, id);
+          this._idToObject.set(id, new WeakRef(value));
+        }
+      } else {
+        throw new Error('not yet implemented');
+      }
+      return { type: 'object', objectId: uuid() };
+    }
+
+    deserialize(value) {
+      switch (value.type) {
+        case 'undefined': {
+          return undefined;
+        }
+        case 'null': {
+          return null;
+        }
+        case 'string': {
+          return value.value;
+        }
+        case 'number': {
+          if (value.value === 'NaN') {
+            return NaN;
+          } else if (value.value === '-0') {
+            return -0;
+          } else if (value.value === '+Infinity') {
+            return +Infinity;
+          } else if (value.value === '-Infinity') {
+            return -Infinity;
+          } else {
+            return value.value;
+          }
+        }
+        case 'boolean': {
+          return value.value;
+        }
+        case 'object': {
+          const weakRef = this._idToObject.get(value.objectId);
+          if (!weakRef) {
+            throw new Error('unknown object reference');
+          }
+
+          const obj = weakRef.deref();
+          if (!obj) {
+            throw new Error('stable object reference');
+          }
+
+          return obj;
+        }
+        default:
+          throw new Error('not yet implemented');
+      }
+    }
+  }
+
+  function uuid() {
+    // TODO: Generate UUIDs.
+    return '';
+  }
+
+  return function evaluate(script, args) {
+    if (window.__webdriver_js_serializer === undefined) {
```

Comment:
What if some malicious page change the serialisation function? They could mimic the eval results to anything and make BiDi working unpredictably/unstable on such pages.

Reply:
Agreed this is not ideal. Leaving it as-is for now per our meeting discussions and can look at more robust solutions such as updating CDP in the future. This will at least provide us will spec-compliant values for well-behaved pages which will be useful for testing.